clang/lib/CIR/Dialect/Transforms/LoweringPrepareArm64CXXABI.cpp

```diff
@@ -0,0 +1,127 @@
+//====- LoweringPrepareArm64CXXABI.cpp - Arm64 ABI specific code --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides ARM64 C++ ABI specific code that is used during LLVMIR
+// lowering prepare.
+//
+//===----------------------------------------------------------------------===//
+
+#include "LoweringPrepareItaniumCXXABI.h"
+
+using cir::LoweringPrepareCXXABI;
+using namespace mlir;
+using namespace mlir::cir;
+
+namespace {
+class LoweringPrepareArm64CXXABI : public LoweringPrepareItaniumCXXABI {
+public:
+  mlir::Value lowerVAArg(CIRBaseBuilderTy &builder,
+                         mlir::cir::VAArgOp op) override;
+};
+} // namespace
+
+LoweringPrepareCXXABI *LoweringPrepareCXXABI::createArm64ABI() {
+  return new LoweringPrepareArm64CXXABI();
+}
+
+static mlir::Value createGetMemberOp(CIRBaseBuilderTy &builder,
+                                     mlir::Location &loc, mlir::Value structPtr,
+                                     const char *fldName, unsigned idx) {
+
+  assert(structPtr.getType().isa<mlir::cir::PointerType>());
+  auto structBaseTy =
+      structPtr.getType().cast<mlir::cir::PointerType>().getPointee();
+  assert(structBaseTy.isa<mlir::cir::StructType>());
+  auto fldTy = structBaseTy.cast<mlir::cir::StructType>().getMembers()[idx];
+  auto fldPtrTy = ::mlir::cir::PointerType::get(builder.getContext(), fldTy);
+  return builder.create<mlir::cir::GetMemberOp>(loc, fldPtrTy, structPtr,
+                                                fldName, idx);
+}
```

Comment:
Is this function general or is it specific to ARM64 ABI? If it is general you can move it to `CIRBaseBuilderTy`.

Reply:
This is general, will move to CIRBaseBuilderTy